odd-platform-api/src/main/java/org/opendatadiscovery/oddplatform/service/LineageServiceImpl.java

```diff
@@ -41,91 +42,103 @@
-                final Set<String> oddrnsToFetch = relations.stream()
-                    .flatMap(r -> Stream.of(r.getParentOddrn(), r.getChildOddrn()))
-                    .collect(Collectors.toSet());
-                final Map<String, List<String>> groupRelations =
-                    groupEntityRelationRepository.fetchGroupRelations(oddrnsToFetch);
-
-                final Map<String, Integer> childrenCountMap = lineageRepository.getChildrenCount(oddrnsToFetch);
-                final Map<String, Integer> parentsCountMap = lineageRepository.getParentCount(oddrnsToFetch);
-                final Map<String, DataEntityDimensionsDto> dataEntityDimensionsDtoMap =
-                    dataEntityRepository.listDimensionsByOddrns(SetUtils.union(oddrnsToFetch, groupRelations.keySet()))
-                        .stream()
-                        .collect(Collectors.toMap(d -> d.getDataEntity().getOddrn(), identity()));
-                final Map<DataEntityDimensionsDto, List<String>> groupRepository = groupRelations.entrySet()
-                    .stream()
-                    .map(e -> {
-                        final DataEntityDimensionsDto groupDto = dataEntityDimensionsDtoMap.get(e.getKey());
-                        if (groupDto == null) {
-                            return null;
-                        }
-                        return Pair.of(groupDto, e.getValue());
-                    })
-                    .filter(Objects::nonNull)
-                    .collect(Collectors.toMap(Pair::getLeft, Pair::getRight));
-                final DataEntityLineageStreamDto lineageStream = getLineageStream(dataEntityDimensionsDtoMap,
-                    groupRepository, relations, childrenCountMap, parentsCountMap);
-
-                final var builder = DataEntityLineageDto.builder()
-                    .dataEntityDto(dto);
-                if (lineageStreamKind == DOWNSTREAM) {
-                    return builder.downstream(lineageStream)
-                        .build();
-                } else {
-                    return builder.upstream(lineageStream)
-                        .build();
-                }
-            })
+            .flatMap(dto -> lineageRepository
+                .getLineageRelations(Set.of(dto.getDataEntity().getOddrn()),
+                    LineageDepth.of(lineageDepth), lineageStreamKind)
+                .collectList()
+                .flatMap(relations -> {
+                    final Set<String> oddrnsToFetch = relations.stream()
+                        .flatMap(r -> Stream.of(r.getParentOddrn(), r.getChildOddrn()))
+                        .collect(Collectors.toSet());
+                    return Mono
+                        .zip(groupEntityRelationRepository.fetchGroupRelations(oddrnsToFetch),
+                            lineageRepository.getChildrenCount(oddrnsToFetch),
+                            lineageRepository.getParentCount(oddrnsToFetch))
+                        .map(tuple -> {
```

Comment:
you can use TupleUtils.finction() for giving names to tuples components. There are a lot of examples in codebase